chain/checkpointing/sub.go

```diff
@@ -0,0 +1,506 @@
+	config *keygen.TaprootConfig
+	// new config generated
+	newconfig *keygen.TaprootConfig
+	// Initiated
+	init bool
+	// Previous tx
+	ptxid string
+	// Tweaked value
+	tweakedValue []byte
+}
+
+func NewCheckpointSub(
+	mctx helpers.MetricsCtx,
+	lc fx.Lifecycle,
+	host host.Host,
+	pubsub *pubsub.PubSub,
+	api impl.FullNodeAPI,
+) (*CheckpointingSub, error) {
+
+	ctx := helpers.LifecycleCtx(mctx, lc)
+	// Starting shardSub to listen to events in the root chain.
+	e, err := events.NewEvents(ctx, &api)
+	if err != nil {
+		return nil, err
+	}
+
+	var config *keygen.TaprootConfig
+	// Load configTaproot
+	if _, err := os.Stat(os.Getenv("EUDICO_PATH") + "/share.toml"); errors.Is(err, os.ErrNotExist) {
+		// path/to/whatever does not exist
+		fmt.Println("No share file saved")
+	} else {
+		content, err := os.ReadFile(os.Getenv("EUDICO_PATH") + "/share.toml")
+		if err != nil {
+			return nil, err
+		}
+
+		var configTOML TaprootConfigTOML
+		_, err = toml.Decode(string(content), &configTOML)
+		if err != nil {
+			return nil, err
+		}
+
+		privateSharePath, err := hex.DecodeString(configTOML.PrivateShare)
+		if err != nil {
+			return nil, err
+		}
+
+		publickey, err := hex.DecodeString(configTOML.PublicKey)
+		if err != nil {
+			return nil, err
+		}
+
+		var privateShare curve.Secp256k1Scalar
+		err = privateShare.UnmarshalBinary(privateSharePath)
+		if err != nil {
+			return nil, err
+		}
+
+		verificationShares := make(map[party.ID]*curve.Secp256k1Point)
+
+		fmt.Println(configTOML.VerificationShares)
+
+		for key, vshare := range configTOML.VerificationShares {
+
+			fmt.Println(key)
+			fmt.Println(vshare)
+
+			var p curve.Secp256k1Point
+			pByte, err := hex.DecodeString(vshare.Share)
+			if err != nil {
+				return nil, err
+			}
+			err = p.UnmarshalBinary(pByte)
+			if err != nil {
+				return nil, err
+			}
+			verificationShares[party.ID(key)] = &p
+		}
+
+		config = &keygen.TaprootConfig{
+			ID:                 party.ID(host.ID().String()),
+			Threshold:          configTOML.Thershold,
+			PrivateShare:       &privateShare,
+			PublicKey:          publickey,
+			VerificationShares: verificationShares,
+		}
+	}
+
+	return &CheckpointingSub{
+		pubsub:    pubsub,
+		topic:     nil,
+		sub:       nil,
+		host:      host,
+		api:       &api,
+		events:    e,
+		init:      false,
+		ptxid:     "",
+		config:    config,
+		newconfig: nil,
+	}, nil
+}
+
+func (c *CheckpointingSub) listenCheckpointEvents(ctx context.Context) {
+
+	checkFunc := func(ctx context.Context, ts *types.TipSet) (done bool, more bool, err error) {
+		return false, true, nil
+	}
+
+	changeHandler := func(oldTs, newTs *types.TipSet, states events.StateChange, curH abi.ChainEpoch) (more bool, err error) {
+		log.Infow("State change detected for power actor")
+
+		idsStrings := c.orderParticipantsList()
+
+		fmt.Println("Participants list :", idsStrings)
+
+		ids := c.formIDSlice(idsStrings)
+
+		id := party.ID(c.host.ID().String())
+
+		threshold := 2
```

Comment:
Why is the threshold 2? Are we assuming three nodes here? We should probably adapt this according to the number of miners in the network, right?